### PR TITLE
Enh/container flag new

### PR DESCRIPTION
**Pull request by djarecka**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Summary
this replaces #183

adding `container` argument to the `ShellCommandTask`, so it can run in a container 

## Checklist

- [x] All tests passing
- [x] I have added tests to cover my changes
- [ ] I have updated documentation (if necessary)
- [x] My code follows the code style of this project 
(we are using `black`: you can `pip install pre-commit`, 
run `pre-commit install` in the `pydra` directory 
and `black` will be run automatically with each commit)

## Acknowledgment
- [x] I acknowledge that this contribution will be available under the Apache 2 license.

### CODE REVIEW

**Comment by djarecka** [comment]
@satra - you're right, that with `__new__` is much better, but I failed to use `container_info` as you suggested, had to provide `image` as a separate argument. There has been something important I'm missing about `__new__`...

**Comment by satra** [comment]
@djarecka - perhaps this demo can help

```python
In [10]: class A: 
    ...:     def __new__(cls, *args, **kwargs): 
    ...:         if 'info' in kwargs: 
    ...:             print('Info: {0}'.format(kwargs['info'])) 
    ...:         return super(A, cls).__new__(cls) 
    ...:     def __init__(self, info=None): 
    ...:         pass 
    ...:                                                                                       

In [11]: a = A()                                                                               

In [12]: a = A(info=('a', 'b'))                                                                
Info: ('a', 'b')

In [13]: a = A(info=None)                                                                      
Info: None
```

so in the shell task info could be a dictionary.

**Comment by djarecka** [comment]
@satra - no, this doesn't help. I modified kwargs, to have image etc., but somehow it goes twice to `DockerTask.__init__` and the second time there is no `image` in the `input`.

**Comment by djarecka** [comment]
I was trying something like [this](https://github.com/djarecka/pydra/blob/enh/container_flag_new_tmp/pydra/engine/task.py#L213), but it doesn't work.

You can try to run: pydra/engine/tests/test_dockertask.py::test_docker_1_dockerflag

**Comment by satra** [comment]
> somehow it goes twice to DockerTask.__init__

that's because dockertask is a subclass of shelltask.

```python
In [29]: class A: 
    ...:     def __new__(cls, *args, **kwargs): 
    ...:         if 'info' in kwargs: 
    ...:             print('Info: {0}'.format(kwargs['info'])) 
    ...:             instance =  B(image=kwargs['info']['image']) 
    ...:             print(type(instance)) 
    ...:             return instance 
    ...:         return super(A, cls).__new__(cls) 
    ...:     def __init__(self, info=None): 
    ...:         print('A init called') 
    ...: class B(A): 
    ...:     def __init__(self, image=None, **kwargs): 
    ...:         print('B entry') 
    ...:         self.image = image 
    ...:         print(kwargs) 
    ...:         print('B exit') 
    ...:                                                                                       

In [30]: a = A(info={'image': 'test'})                                                         
Info: {'image': 'test'}
B entry
{}
B exit
<class '__main__.B'>
B entry
{'info': {'image': 'test'}}
B exit
```

as you can see first the instance of B is created which calls init, and then since new is called inside the original A, another init is invoked with the original arguments. you could avoid it by trapping to see if the instance has already been initialized. of course in the proper implementation you would want to pass the full set of args/kwargs (except info) to B

```python
In [31]: class A: 
    ...:     def __new__(cls, *args, **kwargs): 
    ...:         if 'info' in kwargs: 
    ...:             print('Info: {0}'.format(kwargs['info'])) 
    ...:             instance =  B(image=kwargs['info']['image']) 
    ...:             print(type(instance)) 
    ...:             return instance 
    ...:         return super(A, cls).__new__(cls) 
    ...:     def __init__(self, info=None): 
    ...:         print('A init called') 
    ...: class B(A): 
    ...:     init = False 
    ...:     def __init__(self, image=None, **kwargs): 
    ...:         print('B entry') 
    ...:         if not self.init: 
    ...:             self.image = image 
    ...:             print(kwargs) 
    ...:             self.init = True 
    ...:         print('B exit') 
    ...:                                                                                       

In [32]: a = A(info={'image': 'test'})                                                         
Info: {'image': 'test'}
B entry
{}
B exit
<class '__main__.B'>
B entry
B exit
```

**Comment by djarecka** [comment]
@satra i think I tried similar things, but I will give it another try, but not today. thanks

**Comment by djarecka** [comment]
@satra - on sat didn't read your answer carefully, indeed adding init to DockerTask can solve the problem, thanks!

**Comment by djarecka** [comment]
@satra - can i merge it?

File: pydra/engine/task.py [hunk 1]
```diff
@@ -210,6 +210,29 @@ def _run_task(self):
 class ShellCommandTask(TaskBase):
     """Wrap a shell command as a task element."""
 
+    def __new__(cls, container_info=None, *args, **kwargs):
+        if not container_info:
+            return super(ShellCommandTask, cls).__new__(cls)
+
+        if len(container_info) == 3:
+            type, image, bind = container_info
```
Comment (satra) [inline pydra/engine/task.py]: perhaps not use `type` which is a python keyword.

File: pydra/engine/task.py [hunk 2]
```diff
@@ -319,21 +342,28 @@ def command_args(self):
     def command_args(self, args: ty.Dict):
         self.inputs = attr.evolve(self.inputs, **args)
 
+    @property
+    def container_args(self):
```
Comment (satra) [inline pydra/engine/task.py]: is this still required.

File: pydra/engine/task.py [hunk 3]
```diff
@@ -319,21 +342,28 @@ def command_args(self):
...
+
     @property
     def cmdline(self):
         """Get the actual command line that will be submitted."""
         orig_inputs = attr.asdict(self.inputs)
         modified_inputs = template_update(self.inputs)
         if modified_inputs is not None:
             self.inputs = attr.evolve(self.inputs, **modified_inputs)
-        cmdline = " ".join(self.command_args)
+        cmdline = " ".join(self.container_args + self.command_args)
```
Comment (satra) [inline pydra/engine/task.py]: should this and the container cmdline be reverted.

File: pydra/engine/task.py (hunk 2 above)
Comment (djarecka) [reply pydra/engine/task.py]: I added this when I decided to remove `run_task` and `smndline` from `ContainerTask`. They were very similar

Comment (djarecka) [reply pydra/engine/task.py]: but i can modify the methods and add an if statement

File: pydra/engine/task.py (hunk 3 above)
Comment (djarecka) [reply pydra/engine/task.py]: not sure if i follow, looks good to me

**Comment by satra** [comment]
lgtm - you can merge once tests pass.

**Pull request merged**